src/proof/arith_proof.cpp

```diff
@@ -24,15 +24,84 @@
 #include "proof/theory_proof.h"
 #include "theory/arith/constraint_forward.h"
 #include "theory/arith/theory_arith.h"
+#include "theory/rewriter.h"
 
 #define CVC4_ARITH_VAR_TERM_PREFIX "term."
 
 namespace CVC4 {
+using namespace CVC4::theory;
 
 inline static Node eqNode(TNode n1, TNode n2) {
   return NodeManager::currentNM()->mkNode(kind::EQUAL, n1, n2);
 }
 
+// Proofs are supported only for contraints of the form __ <= __ .
+inline static bool allChildrenGeq(const Node& conflict)
+{
+  bool result = true;
+  size_t nChildren = conflict.getNumChildren();
+  for (size_t i = 0; i < nChildren; i++)
+  {
+    const Node& n = conflict[i];
+    const Node& nonneg = n.getKind() == kind::NOT ? n[0] : n;
+    if (nonneg.getKind() == kind::GEQ)
+    {
+      continue;
+    }
+    else
+    {
+      result = false;
+      break;
+    }
+  }
+  return result;
+}
+
+// Verify that that the farkas coefficients indeed create a contradiction
+inline static bool hasContradiction(
+    const Node& conflict, theory::arith::RationalVectorCP farkasCoefficients)
+{
+  NodeManager* nm = NodeManager::currentNM();
+  NodeBuilder<> leftBuilder(kind::PLUS);
+  NodeBuilder<> rightBuilder(kind::PLUS);
+  Node currentLeft;
+  Node currentRight;
+  bool hasStrictIneq = false;
+  // add all inequalities after they are multiplied with their Farkas
+  // coefficients
+  const size_t n = conflict.getNumChildren();
+  for (size_t i = 0; i != n; i++)
+  {
+    const Node& lem = conflict[i];
+    Rational c = (*farkasCoefficients)[i].abs();
+    // if lem has the form ~(x <= y), this is equivalent to x > y.
+    // Hence the Farkas coefficient is multiplied by -1, to keep the inequality
+    // in the right direction. Also, the resulting inequality will be strict.
+    if (lem.getKind() == kind::NOT)
+    {
+      hasStrictIneq = true;
+      currentLeft = lem[0][0];
+      Assert(lem[0].getKind() == kind::GEQ);
```

Comment:
It would be more natural for me to move this assertion up by a line.

Reply:
fixed in next commit.